File: outbox-pattern-practice/src/main/java/com/example/outboxPatternPractice/service/OrderService.java
package com.example.outboxPatternPractice.service;

import com.example.outboxPatternPractice.entity.EventOutbox;
import com.example.outboxPatternPractice.entity.Order;
import com.example.outboxPatternPractice.entity.OrderDetail;
import com.example.outboxPatternPractice.repository.EventOutboxRepository;
import com.example.outboxPatternPractice.repository.OrderDetailRepository;
import com.example.outboxPatternPractice.repository.OrderRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.UUID;

@Service
public class OrderService {

    private final OrderRepository orderRepository;
    private final OrderDetailRepository orderDetailRepository;
    private final EventOutboxRepository eventOutboxRepository;

    public OrderService(OrderRepository orderRepository, OrderDetailRepository orderDetailRepository, EventOutboxRepository eventOutboxRepository) {
        this.orderRepository = orderRepository;
        this.orderDetailRepository = orderDetailRepository;
        this.eventOutboxRepository = eventOutboxRepository;
    }

    @Transactional
    public void create(Long productId, Long userId, int quantity) {
        LocalDateTime now = LocalDateTime.now();
        UUID uuid = UUID.randomUUID();

        orderRepository.save(new Order(uuid, userId, now));
        orderDetailRepository.save(new OrderDetail(productId, quantity, now));
        eventOutboxRepository.save(new EventOutbox(uuid));
    }
}

File: loop-pack-be-l2-vol3-kotlin/apps/commerce-api/src/main/kotlin/com/loopers/domain/order/OrderService.kt
package com.loopers.domain.order

import com.loopers.domain.support.PageResult
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

/**
 * 주문 애그리거트의 유스케이스.
 *
 * 이 서비스는 상품도 회원도 모른다. 재고를 차감하는 것은 이 애그리거트의 일이 아니며,
 * 세 애그리거트를 잇는 책임은 OrderFacade 에만 있다. (설계 문서 7.2 장)
 *
 * 그래서 place 가 받는 것이 productId 가 아니라 이미 스냅샷이 채워진 OrderItemModel 이다.
 * 상품을 조회해 이름과 가격을 읽는 일은 호출자가 끝내고 들어온다.
 */
@Component
class OrderService(
    private val orderRepository: OrderRepository,
) {
    @Transactional
    fun place(userId: Long, items: List<OrderItemModel>): OrderModel {
        return orderRepository.save(OrderModel.create(userId = userId, items = items))
    }

    /** 없으면 null 이다. 남의 주문인지 판정하는 것은 이 계층의 일이 아니다. */
    @Transactional(readOnly = true)
    fun getOrder(id: Long): OrderModel? {
        return orderRepository.findById(id)
    }

    @Transactional(readOnly = true)
    fun getOrders(criteria: OrderCriteria.Search): PageResult<OrderModel> {
        return orderRepository.findAll(criteria)
    }
}
